File: src/main/java/com/walshtransform/NKLandscapeWalshTransformer.java
package com.walshtransform;

import java.util.HashMap;
import java.util.Map;

/**
 * Grey-box Walsh coefficient extractor for NK landscapes.
 */
public final class NKLandscapeWalshTransformer {

    private NKLandscapeWalshTransformer() {
        // Utility class; prevent instantiation.
    }

    /**
     * Extracts sparse Walsh coefficients by exploiting NK landscape structure.
     */
    public static Map<Long, Double> extractCoefficients(NKLandscape nk) {
        if (nk == null) {
            throw new NullPointerException("nk must not be null");
        }
        int n = nk.getDimension();
        if (n >= Long.SIZE) {
            throw new IllegalArgumentException("n must be < 64 to fit in a long mask");
        }

        int k = nk.getK();
        int localSize = 1 << (k + 1); // Size of each local truth table: 2^(k+1)
        double[][][] tables = nk.getInteractionTables();
        int[][] interactions = nk.getInteractions();

        // Sparse map of global Walsh coefficients: mask -> coefficient value.
        Map<Long, Double> coefficients = new HashMap<>(n * localSize);
        double localScale = 1.0 / localSize; // FWHT normalization for local tables

        for (int i = 0; i < n; i++) {
            // Copy the local contribution table for f_i into a flat array.
            double[] local = new double[localSize];
            for (int idx = 0; idx < localSize; idx++) {
                local[idx] = tables[i][idx][0];
            }

            // Transform local values into local Walsh coefficients in-place.
            FastWalshTransformer.fwht(local);

            for (int idx = 0; idx < localSize; idx++) {
                // Normalize local Walsh coefficient for this local mask.
                double value = local[idx] * localScale;
                if (Math.abs(value) < 1e-12) {
                    continue;
                }

                // Map local mask bits to their global variable positions.
                long globalMask = 0L;
                int localMask = idx;
                for (int bit = 0; bit <= k; bit++) {
                    if ((localMask & 1) != 0) {
                        globalMask |= 1L << interactions[i][bit];
                    }
                    localMask >>= 1;
                }

                // Accumulate into the sparse global coefficient map.
                coefficients.merge(globalMask, value, Double::sum);
            }
        }

        // NK fitness is the average of n local contributions.
        double globalScale = 1.0 / n;
        for (Map.Entry<Long, Double> entry : coefficients.entrySet()) {
            entry.setValue(entry.getValue() * globalScale);
        }

        return coefficients;
    }
}

File: python/walshtransform/nk_landscape_walsh_transformer.py
from typing import Dict, FrozenSet
from .nk_landscape import NKLandscape
from .fast_walsh_transformer import FastWalshTransformer

class NKLandscapeWalshTransformer:
    """
    Grey-box Walsh coefficient extractor for NK landscapes exploiting sparsity.
    """
    @staticmethod
    def extract_coefficients(nk: NKLandscape) -> Dict[FrozenSet[int], float]:
        if nk is None:
            raise ValueError("nk must not be None")
            
        n = nk.dimension
        k = nk.k
        local_size = 1 << (k + 1)
        tables = nk.interaction_tables
        interactions = nk.interactions
        
        coefficients = {}
        local_scale = 1.0 / local_size
        
        for i in range(n):
            # Copy local contribution table
            local = [tables[i][idx][0] for idx in range(local_size)]
            
            # Compute FWHT in-place on local contribution table
            FastWalshTransformer.fwht(local)
            
            for idx in range(local_size):
                value = local[idx] * local_scale
                if abs(value) < 1e-12:
                    continue
                    
                # Map local mask bits to their global variable positions
                global_indices = []
                local_mask = idx
                for bit in range(k + 1):
                    if (local_mask & 1) != 0:
                        global_indices.append(interactions[i][bit])
                    local_mask >>= 1
                    
                key = frozenset(global_indices)
                coefficients[key] = coefficients.get(key, 0.0) + value
                
        # Scale by 1/n
        global_scale = 1.0 / n
        for key in list(coefficients.keys()):
            coefficients[key] *= global_scale
            
        return coefficients
